contracts/settling_game/modules/crypts/CryptRun.cairo

```diff
@@ -0,0 +1,277 @@
+    new_implementation: felt
+) {
+    Proxy.assert_only_admin();
+    Proxy._set_implementation_hash(new_implementation);
+    return ();
+}
+
+// -----------------------------------
+// EXTERNAL
+// -----------------------------------
+
+// @notice Opens crypt.... Checks if the Crypt is active. If < 24hrs or > 10 successful quests, function creates a new seed.
+// @param crypt_id: Crypts erc721 ID
+// @param adventurer_id: Adventurer erc721 ID
+// @returns seed: Random seed. This is used in the Client to reconstruct the Crypt in a visual form.
+@external
+func open_crypt{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    crypt_id: Uint256, adventurer_id: Uint256
+) -> (seed: felt) {
+    // check ownership
+    assert_adventurer(adventurer_id);
+
+    let (now) = get_block_timestamp();
+
+    let (crypt_status) = crypt.read(crypt_id);
+    let is_seed_active = is_le(now, crypt_status.time_stamp);
+
+    if (is_seed_active == FALSE) {
+        // get random
+        tempvar randomSeed = 123456842657854254785;
+
+        // write crypt set ts 1 day from now
+        crypt.write(crypt_id, Crypt(now + DAY, randomSeed, NUMBER_SUCCESSFUL_RUNS));
+
+        // write adventurer
+        adventurer.write(
+            adventurer_id, AdventurerLocation(crypt_id.low, NODE_START_INDEX, randomSeed)
+        );
+
+        return (seed=randomSeed);
+    }
+
+    // we save the seed within the adventurer struct, this is incase the seed changes midrun the player can still complete
+    adventurer.write(
+        adventurer_id, AdventurerLocation(crypt_id.low, NODE_START_INDEX, crypt_status.seed)
+    );
+
+    return (seed=crypt_status.seed);
+}
+
+// @notice Move Adventurer within the Crypt.
+// @param crypt_id: Crypts erc721 ID
+// @param adventurer_id: Adventurer erc721 ID
+// @param start_location: Index of the current location
+// @param end_location: Index of the location wishing to travel to
+@external
+func move_to_location_in_crypt{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    crypt_id: Uint256, adventurer_id: Uint256, start_location: felt, end_location: felt
+) -> () {
+    alloc_locals;
+    // check ownership
+    assert_adventurer(adventurer_id);
+
+    // assert at location requested
+    let (adventurer_location) = assert_and_return_location(crypt_id, adventurer_id, start_location);
+
+    let (dungeon) = build_dungeon(crypt_id, adventurer_location.seed);
+
+    // check path exists
+    let (shortest_path_len, identifiers, total_distance) = Crypts.check_path_exists(
+        dungeon, start_location, end_location
+    );
+
+    // check path exists
+    with_attr error_message("Crypts: unmoveable location from your location") {
+        assert_not_zero(shortest_path_len);
+    }
+
+    // move player and store state
+    adventurer.write(
+        adventurer_id, AdventurerLocation(crypt_id.low, end_location, adventurer_location.seed)
```

Comment:
So it's possible to move through multiple vertices in one function call? Say in the [example](https://dreampuf.github.io/GraphvizOnline/#digraph%20G%20%7B%0A0%20-%3E%20%7B1%20%7D%0A1%20-%3E%20%7B2%20100%20%7D%0A2%20-%3E%20%7B3%20200%20%7D%0A3%20-%3E%20%7B4%20%7D%0A4%20-%3E%20%7B5%20400%20%7D%0A5%20-%3E%20%7B6%20500%20%7D%0A6%20-%3E%20%7B7%20%7D%0A7%20-%3E%20%7B8%20700%20%7D%0A8%20-%3E%20%7B9%20%7D%0A9%20-%3E%20%7B10%20%7D%0A10%20-%3E%20%7B1000%20%7D%0A500%20-%3E%20%7B6%20%7D%0A200%20-%3E%20%7B6%20%7D%0A1000%20-%3E%20%7B6%20%7D%0A700%20-%3E%20%7B6%20%7D%0A400%20-%3E%20%7B6%20%7D%0A100%20-%3E%20%7B6%20%7D%0A%7D), if I'm at `2`, I can move to `700`? Intuitively, I'd assume it's only possible to move to adjacent nodes (i.e. `3` or `200` or `1`).

Reply:
you can move IF a path exists between the nodes and there is no entity blocking your path. The Dijkstra checks this, and if a path exists returns the length. If the length is 0, then you cannot move to the Vertex you are trying to get to.

Reply:
I understand that from an algo point of view. My comment was more about the gameplay - should this be allowed, skipping through rooms (that's my mental model, correct me if I'm wrong).

Reply:
One other thing - you mentioned dungeons are directed graphs. Does the Dijkstra implementation take that into account? Or would it return a non zero when called with 700 as `start_location` and 2 as `end_location`?

Reply:
unless we wanted to put an interaction in each room, I think we should have the option to traverse many rooms at once. But open to alternatives.

Correct -Dijkstra would return 0 it cannot go backward.